### PR TITLE
Improve re-throwing exceptions by using ExceptionDispatchInfo throughout (Issue #2445)

### PR DESCRIPTION
**Pull request by macrogreg**
Addresses https://github.com/Azure/azure-webjobs-sdk/issues/2445
Please see the description there.

### CODE REVIEW

**Comment by brettsam** [comment]
For external PRs we need to manually add the `full-ci` label to get CI to run correctly... I've just added it.

**Comment by brettsam** [comment]
@mathewc / @fabiocav -- this seems like general goodness, especially in some of our more complicated `catch` blocks. I think we do some fiddling with stack traces (in Functions?) to make errors look nicer -- any worries that this may affect that? I don't think there is but want to call it out.

File: src/Microsoft.Azure.WebJobs.Extensions.Storage/Blobs/Bindings/StorageBlobContainerExtensions.cs [hunk 1]
```diff
@@ -48,7 +49,8 @@ internal static class StorageBlobContainerExtensions
 
                 if (result == null || result.HttpStatusCode != 404)
                 {
-                    throw;
+                    ExceptionDispatchInfo.Capture(exception).Throw();
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.Storage/Blobs/Bindings/StorageBlobContainerExtensions.cs:52]: Here and in all the other instances below you're doing an explicit recapture/throw for exceptions that we're already using the "throw" keyword for. My understanding is that as long as we're not doing a rethrow of the form "throw ex", the stack trace is maintained already without a recapture.

It seems we're already doing this correctly in most places. E.g. in places in our execution pipeline where we catch an invocation exception and don't throw immediately, we're capturing and throwing the ExceptionDispatchInfo and using it to rethrow later, e.g. [here](https://github.com/Azure/azure-webjobs-sdk/blob/f7da27d133b9ce0aadf9918770f6ed45748f3c8a/src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs#L134).

Comment (macrogreg) [reply src/Microsoft.Azure.WebJobs.Extensions.Storage/Blobs/Bindings/StorageBlobContainerExtensions.cs:52]: Thanks for your response, @mathewc .  This statement is not quite correct: 

> as long as we're not doing a rethrow of the form "throw ex", the stack trace is maintained already without a recapture.

Please see my example in the issue https://github.com/Azure/azure-webjobs-sdk/issues/2445 .
The stack you get with `throw` shows the line where the exception came up in the `try`-block, bit not the line where it was re-thrown in the `catch`-block. The stack you get with `ExceptionDispatchInfo` shows both. Please see the above-linked post for an example.

**Comment by azfuncgh** [comment]
In an effort to clean up the repository, we're closing stale PRs. If you feel this is still relevant, please reopen the PR and we'll work on moving this forward. Thank you for your contribution!

**Pull request closed without merging**